File: iOS/Mobile/CourseDetailViewController.m
//
//  CourseDetailViewController.m
//  Mobile
//
//  Created by Jason Hocker on 9/26/12.
//  Copyright (c) 2012-2014 Ellucian. All rights reserved.
//

#import "CourseDetailViewController.h"
#import "CurrentUser.h"
#import "CourseDetail.h"
#import "CourseDetailInstructor.h"
#import "CourseMeetingPattern.h"
#import "POIDetailViewController.h"
#import "DirectoryViewController.h"
#import "CourseDetailTabBarController.h"
#import "UIViewController+GoogleAnalyticsTrackerSupport.h"
#import "AppearanceChanger.h"
#import "UIScrollView+Size.h"

@interface CourseDetailViewController ()
@property (strong, nonatomic) NSOrderedSet *meetingPatterns;
@property (strong, nonatomic) NSOrderedSet *instructors;
@property (nonatomic, strong) NSDateFormatter *dateFormatter;
@property (nonatomic, strong) NSDateFormatter *timeFormatter;
@property (nonatomic, strong) NSDateFormatter *displayDateFormatter;
@property (nonatomic, strong) NSDateFormatter *displayTimeFormatter;
@property (nonatomic, strong) NSDateFormatter *sisTimeFormatterWTz;
@end

@implementation CourseDetailViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.navigationController.navigationBar.translucent = NO;
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(setupData) name:kCourseDetailInformationLoaded object:nil];
    
    self.sectionTitleLabel.text = nil;
    self.dateLabel.text = nil;
    
    self.facultyLabel.textColor = [UIColor subheaderTextColor];
    self.sectionTitleLabel.textColor = [UIColor subheaderTextColor];
    self.dateLabel.textColor = [UIColor subheaderTextColor];
    
    if([AppearanceChanger isRTL]) {
        self.sectionTitleLabel.textAlignment = NSTextAlignmentRight;
    }
    
    self.navigationItem.title = self.courseNameAndSectionNumber;
    
    self.backgroundView.backgroundColor = [UIColor accentColor];
    
    self.facultyLabelViewHeightConstraint.constant = 0;
}

-(void) viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self sendView:@"Course Overview" forModuleNamed:self.module.name];
    self.widthConstraint.constant = [AppearanceChanger sizeInOrientation:self.interfaceOrientation].width;
    self.courseDescriptionTextViewHeightConstraint.constant = [self.courseDescriptionTextView heightOfContent];
}

- (IBAction)dismiss:(id)sender {
     [self dismissViewControllerAnimated:YES completion:nil];
}

-(NSDateFormatter *)dateFormatter
{
    if(_dateFormatter == nil) {
        _dateFormatter = [[NSDateFormatter alloc] init];
        [_dateFormatter setDateFormat:@"yyyy-MM-dd"];
        [_dateFormatter setTimeZone:[NSTimeZone timeZoneWithName:@"UTC"]];
    }
    return _dateFormatter;
}

-(NSDateFormatter *)timeFormatter
{
    if(_timeFormatter == nil) {
        _timeFormatter = [[NSDateFormatter alloc] init];
        [_timeFormatter setDateFormat:@"HH:mm'Z'"];
        [_timeFormatter setTimeZone:[NSTimeZone timeZoneWithName:@"UTC"]];
    }
    return _timeFormatter;
}

-(NSDateFormatter *)displayDateFormatter
{
    if(_displayDateFormatter == nil) {
        _displayDateFormatter = [[NSDateFormatter alloc] init];
        [_displayDateFormatter setDateStyle:NSDateFormatterShortStyle];
        [_displayDateFormatter setTimeStyle:NSDateFormatterNoStyle];
        [_displayDateFormatter setTimeZone:[NSTimeZone timeZoneWithName:@"UTC"]];
    }
    return _displayDateFormatter;
}

-(NSDateFormatter *)displayTimeFormatter
{
    if(_displayTimeFormatter == nil) {
        _displayTimeFormatter = [[NSDateFormatter alloc] init];
        [_displayTimeFormatter setDateStyle:NSDateFormatterNoStyle];
        [_displayTimeFormatter setTimeStyle:NSDateFormatterShortStyle];
    }
    return _displayTimeFormatter;
}

-(void) setupData
{
    NSString *buildingsUrl = [[NSUserDefaults standardUserDefaults] objectForKey:@"urls-map-buildings"];
    
    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"CourseDetail"];
    request.predicate = [NSPredicate predicateWithFormat:@"termId == %@ && sectionId == %@", self.termId, self.sectionId];
    
    NSError *error;
    CourseDetail *courseDetail = [[self.module.managedObjectContext executeFetchRequest:request error:&error] lastObject];
    self.instructors = courseDetail.instructors;
    self.meetingPatterns = courseDetail.meetingPatterns;
    
    //title
    self.sectionTitleLabel.text = courseDetail.sectionTitle;
    
    //date
    if(courseDetail.firstMeetingDate && courseDetail.lastMeetingDate) {
        NSString *dates = [NSString stringWithFormat:NSLocalizedStringWithDefaultValue(@"course first meeting - last meeting", @"Localizable", [NSBundle mainBundle], @"%@ - %@", @"course first meeting - last meeting"), [self.displayDateFormatter stringFromDate:courseDetail.firstMeetingDate], [self.displayDateFormatter stringFromDate:courseDetail.lastMeetingDate]];        self.dateLabel.text = dates;
    }
    
    //meeting patterns
    id lastObject = self.view;
    if([self.meetingPatterns count] > 0) {
        for (int i = 0; i < [self.meetingPatterns count]; i++) {
            
            CourseMeetingPattern *mp = [self.meetingPatterns objectAtIndex:i];
            
            PseudoButtonView *meetingPatternView = [PseudoButtonView new];
            meetingPatternView.tag = 100 + i;
            meetingPatternView.translatesAutoresizingMaskIntoConstraints = NO;
            
            //mp:days
            NSMutableArray *daysOfClass = [[mp.daysOfWeek componentsSeparatedByString:@"," ] mutableCopy];
            NSArray *localizedDays = [self.dateFormatter shortStandaloneWeekdaySymbols];
            for(int i = 0; i < [daysOfClass count]; i++) {
                NSInteger value = [[daysOfClass objectAtIndex:i] intValue] - 1;
                [daysOfClass replaceObjectAtIndex:i withObject:[localizedDays objectAtIndex:value]];
            }
            
            //mp:time
            NSString *days = [NSString stringWithFormat:NSLocalizedStringWithDefaultValue(@"days:", @"Localizable", [NSBundle mainBundle], @"%@: ", @"days:"), [daysOfClass componentsJoinedByString:@", "]];
            NSString *line1;
            if(mp.instructionalMethodCode) {
                line1 = [NSString stringWithFormat:NSLocalizedStringWithDefaultValue(@"days start - end method", @"Localizable", [NSBundle mainBundle], @"%@ %@ - %@ %@", @"days start - end method"), days, [self.displayTimeFormatter stringFromDate: mp.startTime], [self.displayTimeFormatter stringFromDate:mp.endTime], mp.instructionalMethodCode];
            } else {
                line1 = [NSString stringWithFormat:NSLocalizedStringWithDefaultValue(@"days start - end", @"Localizable", [NSBundle mainBundle], @"%@ %@ - %@", @"days start - end"), days, [self.displayTimeFormatter stringFromDate: mp.startTime], [self.displayTimeFormatter stringFromDate:mp.endTime]];
            }
           

            NSMutableAttributedString *attributedLine1 =[[NSMutableAttributedString alloc]initWithString:line1];
            [attributedLine1 addAttribute:NSFontAttributeName value:[UIFont boldSystemFontOfSize:14.0f] range:NSMakeRange(0, [days length])];
            UILabel *line1Label = [UILabel new];
            line1Label.translatesAutoresizingMaskIntoConstraints = NO;
            line1Label.backgroundColor = [UIColor clearColor];
            [line1Label setFont:[UIFont systemFontOfSize:14.0f]];
            [line1Label setTextColor:[UIColor subheaderTextColor]];
            line1Label.minimumScaleFactor = .5f;
            line1Label.adjustsFontSizeToFitWidth = YES;
            line1Label.textAlignment = [AppearanceChanger isRTL] ? NSTextAlignmentRight : NSTextAlignmentLeft;

            line1Label.attributedText = attributedLine1;
            [meetingPatternView addSubview:line1Label];
            
            //mp:first line
            [meetingPatternView addConstraints: [NSLayoutConstraint constraintsWithVisualFormat:@"|-10-[label]-10-|"
                                                                                        options:NSLayoutFormatAlignAllCenterY metrics:nil
                                                                                          views:@{@"label":line1Label}]];
            
            
            //mp:location
            NSString *location = @"";
            if(mp.building && mp.room) {
                location = [NSString stringWithFormat:NSLocalizedString(@"%@, Room %@", @"label - building name, room number"), mp.building, mp.room];
            } else if(mp.building) {
                location = mp.building;
            } else  if(mp.room) {
                location = [NSString stringWithFormat:NSLocalizedString(@"Room %@", @"label - room number"), mp.room];
            }
            UILabel *locationLabel = [UILabel new];
            locationLabel.backgroundColor = [UIColor clearColor];
            [locationLabel setTextColor:[UIColor subheaderTextColor]];
            [locationLabel setFont:[UIFont systemFontOfSize:14.0f]];
            locationLabel.text = location;
            if(buildingsUrl && mp.buildingId) {
                [meetingPatternView setAction:@selector(gotoLocation:) withTarget:self];
                
                NSMutableAttributedString *underlinedString = [[NSMutableAttributedString alloc] initWithString:location];
                [underlinedString addAttribute:NSUnderlineStyleAttributeName
                                         value:[NSNumber numberWithInt:1]
                                         range:(NSRange){0,[location length]}];
                [underlinedString addAttribute:NSForegroundColorAttributeName
                                         value:[UIColor blueColor]
                                         range:(NSRange){0,[location length]}];
                locationLabel.attributedText = underlinedString;
            }
            locationLabel.translatesAutoresizingMaskIntoConstraints = NO;
            [meetingPatternView addSubview:locationLabel];
            
            //mp:vertical alignment
            [meetingPatternView addConstraints:
             [NSLayoutConstraint constraintsWithVisualFormat:@"V:|-8-[days][location]-8-|"
                                                     options:NSLayoutFormatAlignAllLeading metrics:nil
                                                       views:@{@"days":line1Label, @"location" :locationLabel}]];
            [self.meetingPatternsView addSubview:meetingPatternView];
            
            NSString *vf = @"V:[previous][meeting]";
            if([courseDetail.meetingPatterns count] == 1) {
                vf = @"V:|[meeting]|";
            } else if(i == 0) {
                vf = @"V:|[meeting]";
            } else if(i + 1 == [courseDetail.meetingPatterns count]) {
                vf = @"V:[previous][meeting]|";
            }
            [self.meetingPatternsView addConstraints:
             [NSLayoutConstraint constraintsWithVisualFormat:vf
                                                     options:NSLayoutFormatAlignAllLeading metrics:nil
                                                       views:@{@"previous":lastObject, @"meeting" :meetingPatternView}]];
            [self.meetingPatternsView addConstraints:
             [NSLayoutConstraint constraintsWithVisualFormat:@"|[meeting]|"
                                                     options:0 metrics:nil
                                                       views:@{ @"meeting" :meetingPatternView}]];
            lastObject = meetingPatternView;
        }
    } else {
        self.meetingPatternsView.translatesAutoresizingMaskIntoConstraints = NO;
        [self.meetingPatternsView addConstraint:[NSLayoutConstraint constraintWithItem:self.meetingPatternsView
                                                                              attribute:NSLayoutAttributeHeight
                                                                              relatedBy:NSLayoutRelationEqual
                                                                                 toItem:nil
                                                                              attribute:NSLayoutAttributeNotAnAttribute
                                                                             multiplier:1
                                                                               constant:0]];
    
        
    }

    lastObject = self.view;
    for (int i = 0; i < [self.instructors count]; i++) {
        
        [self.facultyLabelView removeConstraint:self.facultyLabelViewHeightConstraint];
        CourseDetailInstructor *instructor = [self.instructors objectAtIndex:i];
        
        PseudoButtonView *facultyView = [PseudoButtonView new];
        facultyView.tag = 200 + i;
        facultyView.translatesAutoresizingMaskIntoConstraints = NO;
        
        //faculty name
        UILabel *facultyLabel = [UILabel new];
        facultyLabel.backgroundColor = [UIColor clearColor];
        facultyLabel.translatesAutoresizingMaskIntoConstraints = NO;
        facultyLabel.text = instructor.formattedName;
        facultyLabel.font = [UIFont systemFontOfSize:16.0f];
        facultyLabel.textAlignment = [AppearanceChanger isRTL] ? NSTextAlignmentRight : NSTextAlignmentLeft;
        NSString *directoryUrl = [[NSUserDefaults standardUserDefaults] objectForKey:@"urls-directory-facultySearch"];
        if(directoryUrl) {
            [facultyView setAction:@selector(gotoFaculty:) withTarget:self];
        }
        
        [facultyView addSubview:facultyLabel];
        [facultyView addConstraint: [NSLayoutConstraint constraintWithItem:facultyLabel
                                     attribute:NSLayoutAttributeCenterY
                                     relatedBy:NSLayoutRelationEqual
                                        toItem:facultyView
                                     attribute:NSLayoutAttributeCenterY
                                    multiplier:1.0
                                      constant:0]];
        [facultyView addConstraint:[NSLayoutConstraint constraintWithItem:facultyView
                                                                             attribute:NSLayoutAttributeHeight
                                                                             relatedBy:NSLayoutRelationEqual
                                                                                toItem:nil
                                                                             attribute:NSLayoutAttributeNotAnAttribute
                                                                            multiplier:1
                                                                              constant:36.0f]];
        [facultyView addConstraints:
         [NSLayoutConstraint constraintsWithVisualFormat:@"|-(10)-[faculty]-(10)-|"
                                                 options:0 metrics:nil
                                                   views:@{@"faculty":facultyLabel}]];
        [self.facultyView addSubview:facultyView];
        
        NSString *vf = @"V:[previous][faculty]";
        if([courseDetail.instructors count] == 1) {
            vf = @"V:|[faculty]|";
        } else if(i == 0) {
            vf = @"V:|[faculty]";
        } else if(i + 1 == [courseDetail.instructors count]) {
            vf = @"V:[previous][faculty]|";
        }
        [self.facultyView addConstraints:
         [NSLayoutConstraint constraintsWithVisualFormat:vf
                                                 options:NSLayoutFormatAlignAllLeading metrics:nil
                                                   views:@{@"previous":lastObject, @"faculty" :facultyView}]];
        [self.facultyView addConstraints:
         [NSLayoutConstraint constraintsWithVisualFormat:@"|[faculty]|"
                                                 options:0 metrics:nil
                                                   views:@{@"faculty":facultyView}]];
        lastObject = facultyView;
        
    }

    //description
    self.courseDescriptionTextView.text = courseDetail.courseDescription;
    self.courseDescriptionTextView.textAlignment = [AppearanceChanger isRTL] ? NSTextAlignmentRight : NSTextAlignmentLeft;
    
    self.widthConstraint.constant = [AppearanceChanger sizeInOrientation:self.interfaceOrientation].width;

    self.courseDescriptionTextViewHeightConstraint.constant = [self.courseDescriptionTextView heightOfContent];
    self.courseDescriptionTextViewBottomConstraint.constant = 0;
}

- (void)gotoLocation:(UIView *)view {
    NSInteger tag = view.tag - 100;
    CourseMeetingPattern *mp = [self.meetingPatterns objectAtIndex:tag];
    [self performSegueWithIdentifier:@"Show Course Location" sender:mp.buildingId];
}

- (void)gotoFaculty:(UIView *)view {
    NSInteger tag = view.tag - 200;
    CourseDetailInstructor *instructor = [self.instructors objectAtIndex:tag];
    [self performSegueWithIdentifier:@"Show Faculty Person" sender:instructor];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    [self resetScrollViewContentOffset];
    if ([[segue identifier] isEqualToString:@"Show Course Location"])
    {
        [self sendEventToTracker1WithCategory:kAnalyticsCategoryUI_Action withAction:kAnalyticsActionList_Select withLabel:@"Map Detail" withValue:nil forModuleNamed:self.module.name];
        POIDetailViewController *vc = (POIDetailViewController *)[segue destinationViewController];
        vc.buildingId = sender;
        vc.module = self.module;
    } else if ([[segue identifier] isEqualToString:@"Show Faculty Person"])
    {
        [self sendEventWithCategory:kAnalyticsCategoryUI_Action withAction:kAnalyticsActionList_Select withLabel:@"Faculty detail" withValue:nil forModuleNamed:self.module.name];
        DirectoryViewController *detailController = [segue destinationViewController];
        CourseDetailInstructor *instructor = (CourseDetailInstructor *)sender;
        detailController.initialQueryString = [NSString stringWithFormat:@"%@ %@", instructor.firstName, instructor.lastName];
        detailController.initialScope = DirectoryViewTypeFaculty;
        detailController.hideStudents = YES;
        detailController.module = self.module;
    }
}

- (void) willRotateToInterfaceOrientation:(UIInterfaceOrientation)toInterfaceOrientation duration:(NSTimeInterval)duration
{
    self.widthConstraint.constant = [AppearanceChanger sizeInOrientation:toInterfaceOrientation].width;
}

-(void) didRotateFromInterfaceOrientation:(UIInterfaceOrientation)fromInterfaceOrientation
{
    self.courseDescriptionTextViewHeightConstraint.constant = [self.courseDescriptionTextView heightOfContent];
    [self resetScrollViewContentOffset];
}

-(void) resetScrollViewContentOffset
{
    [self.courseDescriptionTextView setContentOffset:CGPointZero animated:YES];
    [self.scrollView setContentOffset:CGPointZero animated:YES];
}


@end

File: iOS/Mobile/CourseDetailViewController.swift
//
//  CourseDetailViewController.swift
//  Mobile
//
//  Created by Jason Hocker on 10/26/15.
//  Copyright © 2015 Ellucian Company L.P. and its affiliates. All rights reserved.
//

import Foundation

class CourseDetailViewController: UITableViewController {
    
    var termId : String?
    var sectionId : String?
    var courseNameAndSectionNumber : String?
    var module : Module?
    var courseDetail : CourseDetail?
    
    let dateFormatter : NSDateFormatter = {
        var formatter = NSDateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.timeZone = NSTimeZone(name:"UTC")
        return formatter
    }()
    let timeFormatter : NSDateFormatter = {
        var formatter = NSDateFormatter()
        formatter.dateFormat = "HH:mm'Z'"
        formatter.timeZone = NSTimeZone(name:"UTC")
        return formatter
    }()
    let displayDateFormatter : NSDateFormatter = {
        var formatter = NSDateFormatter()
        formatter.dateStyle = .ShortStyle
        formatter.timeStyle = .NoStyle
        formatter.timeZone = NSTimeZone(name:"UTC")
        return formatter
    }()
    let displayTimeFormatter : NSDateFormatter = {
        var formatter = NSDateFormatter()
        formatter.dateStyle = .NoStyle
        formatter.timeStyle = .ShortStyle
        return formatter
    }()
    var instructors : [CourseDetailInstructor]?
    var meetingPatterns : [CourseMeetingPattern]?
    var buildingsUrl : String?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationController?.navigationBar.translucent = false
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "setupData", name: "CourseDetailInformationLoaded", object: nil)
        self.navigationItem.title = self.courseNameAndSectionNumber!
        
        tableView.estimatedRowHeight = 60
        tableView.rowHeight = UITableViewAutomaticDimension
        tableView.tableFooterView = UIView()
        
    }
    
    override func tableView(tableView: UITableView, estimatedHeightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        switch(indexPath.section) {
        case 0:
            return 60
        case 1:
            return 60
        default:
            return 120;
        }
        
    }
    
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 3
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if let _ = self.courseDetail {
            switch(section) {
            case 0:
                if let count = self.meetingPatterns?.count where count > 0 {
                    return count
                }
                return 1
            case 1:
                if let count = self.instructors?.count where count > 0 {
                    return count
                }
                return 1
            case 2:
                return courseDetail?.courseDescription != nil ? 1 : 0;
            default:
                return 0;
            }
        }
        return 0;
    }
    
    override func tableView(tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        if let _ = self.courseDetail {
            switch(section) {
            case 0:
                return 52
            case 1:
                return 30
            case 2:
                return 0
            default:
                return 1
            }
        }
        return 0
        
    }
    
    override func tableView(tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        
        if let _ = self.courseDetail {
            switch(section) {
            case 0:
                return headerForTitle()
            case 1:
                return headerForFaculty()
            case 2:
                return nil;
            default:
                return nil;
            }
        }
        return nil;
    }
    
    func headerForTitle() -> UIView? {
        let view = UIView(frame: CGRectMake(0, 0, CGRectGetWidth(tableView.frame), 30))
        let label = UILabel(frame: CGRectMake(8,0,CGRectGetWidth(tableView.frame), 30))
        label.translatesAutoresizingMaskIntoConstraints = false
        let label2 = UILabel(frame: CGRectMake(8,30,CGRectGetWidth(tableView.frame), 30))
        label2.translatesAutoresizingMaskIntoConstraints = false
        
        
        label.textColor = UIColor(red: 0.2, green: 0.2, blue: 0.2, alpha: 1.0)
        label2.textColor = UIColor(red: 0.2, green: 0.2, blue: 0.2, alpha: 1.0)
        view.backgroundColor = UIColor(rgba: "#e6e6e6")
        label.font = UIFont.preferredFontForTextStyle(UIFontTextStyleHeadline)
        label2.font = UIFont.preferredFontForTextStyle(UIFontTextStyleSubheadline)
        
        view.addSubview(label)
        view.addSubview(label2)
        
        let viewsDictionary = ["label": label, "label2": label2, "view": view]
        
        // Create and add the vertical constraints
        view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:|-4-[label]-4-[label2]-4-|",
            options: NSLayoutFormatOptions.AlignAllLeading,
            metrics: nil,
            views: viewsDictionary))
        
        // Create and add the horizontal constraints
        view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("|-[label]",
            options: NSLayoutFormatOptions.AlignAllBaseline,
            metrics: nil,
            views: viewsDictionary))
        view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("|-[label2]",
            options: NSLayoutFormatOptions.AlignAllBaseline,
            metrics: nil,
            views: viewsDictionary))
        
        label.text = self.courseDetail!.sectionTitle;
        if let courseDetail = courseDetail, firstMeetingDate = courseDetail.firstMeetingDate, lastMeetingDate = courseDetail.lastMeetingDate {
            let dates = String(format: NSLocalizedString("course first meeting - last meeting", tableName: "Localizable", bundle: NSBundle.mainBundle(), value: "%@ - %@", comment: "course first meeting - last meeting"), self.displayDateFormatter.stringFromDate(firstMeetingDate), self.displayDateFormatter.stringFromDate(lastMeetingDate))
            label2.text = dates
        }
        return view;
    }
    
    func headerForFaculty() -> UIView? {
        let view = UIView(frame: CGRectMake(0, 0, CGRectGetWidth(tableView.frame), 30))
        let label = UILabel(frame: CGRectMake(8,0,CGRectGetWidth(tableView.frame), 30))
        label.translatesAutoresizingMaskIntoConstraints = false
        
        label.textColor = UIColor(red: 0.2, green: 0.2, blue: 0.2, alpha: 1.0)
        view.backgroundColor = UIColor(rgba: "#e6e6e6")
        label.font = UIFont.preferredFontForTextStyle(UIFontTextStyleHeadline)
        
        view.addSubview(label)
        
        let viewsDictionary = ["label": label, "view": view]
        
        // Create and add the vertical constraints
        view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:|-1-[label]-1-|",
            options: NSLayoutFormatOptions.AlignAllBaseline,
            metrics: nil,
            views: viewsDictionary))
        
        // Create and add the horizontal constraints
        view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("|-[label]",
            options: NSLayoutFormatOptions.AlignAllBaseline,
            metrics: nil,
            views: viewsDictionary))
        
        label.text = NSLocalizedString("Faculty", comment:"Faculty label")
        return view;
        
    }
    
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        self.sendView("Course Overview", forModuleNamed: self.module!.name)
    }
    
    @IBAction func dismiss(sender: AnyObject) {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    func setupData() {
        let defaults = AppGroupUtilities.userDefaults()
        self.buildingsUrl = defaults?.stringForKey("urls-map-buildings")
        let request: NSFetchRequest = NSFetchRequest(entityName: "CourseDetail")
        request.predicate = NSPredicate(format: "termId == %@ && sectionId == %@", self.termId!, self.sectionId!)
        
        self.courseDetail = try! self.module?.managedObjectContext?.executeFetchRequest(request).last as? CourseDetail
        if let courseDetail = self.courseDetail {
            self.instructors = courseDetail.instructors.array as? [CourseDetailInstructor]
            self.meetingPatterns = courseDetail.meetingPatterns.array as? [CourseMeetingPattern]
        }
        self.tableView.reloadData()
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell : UITableViewCell
        
        switch indexPath.section {
        case 0:
            if self.meetingPatterns?.count > 0 {
                let meetingPattern = self.meetingPatterns![indexPath.row]
                cell = tableView.dequeueReusableCellWithIdentifier("Course Detail Meeting Pattern Cell", forIndexPath: indexPath) as UITableViewCell
                cell.userInteractionEnabled = true
                let label = cell.viewWithTag(1) as! UILabel
                let label2 = cell.viewWithTag(2) as! UILabel
                
                var daysOfClass = meetingPattern.daysOfWeek.componentsSeparatedByString(",") as [String]
                var shortStandaloneWeekdaySymbols = self.dateFormatter.shortStandaloneWeekdaySymbols
                var localizedDays = [String]()
                for i in 0 ..< daysOfClass.count {
                    let value = Int(daysOfClass[i])! - 1
                    localizedDays.append(shortStandaloneWeekdaySymbols[value])
                }
                
                //time
                let days: String = String(format: NSLocalizedString("days:", tableName: "Localizable", bundle: NSBundle.mainBundle(), value:"%@: ", comment: "days:"), localizedDays.joinWithSeparator(", "))
                var line1: String
                if meetingPattern.instructionalMethodCode != nil {
                    line1 = String(format: NSLocalizedString("days start - end method", tableName: "Localizable", bundle: NSBundle.mainBundle(), value: "%@ %@ - %@ %@", comment: "days start - end method"), days, self.displayTimeFormatter.stringFromDate(meetingPattern.startTime), self.displayTimeFormatter.stringFromDate(meetingPattern.endTime), meetingPattern.instructionalMethodCode)
                }
                else {
                    line1 = String(format: NSLocalizedString("days start - end", tableName: "Localizable", bundle: NSBundle.mainBundle(), value:"%@ %@ - %@", comment:"days start - end"), days, self.displayTimeFormatter.stringFromDate(meetingPattern.startTime), self.displayTimeFormatter.stringFromDate(meetingPattern.endTime))
                }
                
                let attributedLine1: NSMutableAttributedString = NSMutableAttributedString(string: line1)
                attributedLine1.addAttribute(NSFontAttributeName, value: UIFont.preferredFontForTextStyle(UIFontTextStyleHeadline), range: NSMakeRange(0, days.characters.count))
                label.translatesAutoresizingMaskIntoConstraints = false
                label.attributedText = attributedLine1
                
                //location
                var location: String = ""
                if meetingPattern.building != nil && meetingPattern.room != nil {
                    location = String(format: NSLocalizedString("%@, Room %@", comment: "label - building name, room number"), meetingPattern.building, meetingPattern.room)
                }
                else {
                    if let _ = meetingPattern.building {
                        location = meetingPattern.building
                    }
                    else {
                        if let _ = meetingPattern.room {
                            location = String(format: NSLocalizedString("Room %@", comment: "label - room number"), meetingPattern.room)
                        }
                    }
                }
                if let _ = self.buildingsUrl, _ = meetingPattern.buildingId {
                    
                    let underlineAttributes = [NSUnderlineStyleAttributeName: NSUnderlineStyle.StyleSingle.rawValue,
                        NSForegroundColorAttributeName: UIColor.primaryColor()]
                    let underlineAttributedString = NSAttributedString(string: location, attributes: underlineAttributes)
                    label2.attributedText = underlineAttributedString
                    
                } else {
                    label2.text = location
                }
                
                let imageView = cell.viewWithTag(3)
                imageView!.tintColor = UIColor.primaryColor()
            } else {
                cell = tableView.dequeueReusableCellWithIdentifier("Course Detail No Meeting Pattern Cell", forIndexPath: indexPath) as UITableViewCell
                cell.userInteractionEnabled = false
            }
            
        case 1:
            if self.instructors!.count > 0 {
                let instructor = self.instructors![indexPath.row]
                cell = tableView.dequeueReusableCellWithIdentifier("Course Detail Instructor Cell", forIndexPath: indexPath) as UITableViewCell
                let label = cell.viewWithTag(1) as! UILabel
                label.text = instructor.formattedName
                cell.userInteractionEnabled = true
            } else {
                cell = tableView.dequeueReusableCellWithIdentifier("Course Detail No Instructor Cell", forIndexPath: indexPath) as UITableViewCell
                cell.userInteractionEnabled = false
            }
            
        case 2:
            cell = tableView.dequeueReusableCellWithIdentifier("Course Detail Description Cell", forIndexPath: indexPath) as UITableViewCell
            let label = cell.viewWithTag(1) as! UILabel
            label.text = courseDetail?.courseDescription
            cell.userInteractionEnabled = false
        default:
            cell = UITableViewCell()
        }
        
        cell.layoutIfNeeded()
        
        return cell
    }
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        switch indexPath.section {
        case 0:
            let mp: CourseMeetingPattern = self.meetingPatterns![indexPath.row]
            self.performSegueWithIdentifier("Show Course Location", sender: mp.buildingId)
        case 1:
            
            let instructor = self.instructors![indexPath.row]
            let defaults = AppGroupUtilities.userDefaults()
            var urlString : String?
            
            if ConfigurationManager.doesMobileServerSupportVersion("4.5") {
                urlString = defaults?.stringForKey("urls-directory-baseSearch")
            } else {
                urlString = defaults?.stringForKey("urls-directory-facultySearch")
            }
            
            let name : String
            if instructor.firstName != nil  && instructor.lastName != nil  {
                name = instructor.firstName + " " + instructor.lastName
            } else if instructor.firstName != nil  {
                name = instructor.firstName
            } else if instructor.lastName != nil {
                name = instructor.lastName
            } else {
                name = instructor.formattedName
            }
            
            let encodedSearchString = name.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLQueryAllowedCharacterSet())
            let encodedIdString = instructor.instructorId.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLQueryAllowedCharacterSet())
            urlString = "\(urlString!)?searchString=\(encodedSearchString!)&targetId=\(encodedIdString!)"
            let authenticatedRequest = AuthenticatedRequest()
            let responseData: NSData = authenticatedRequest.requestURL(NSURL(string: urlString!)!, fromView: self)
            let entries = DirectoryEntry.parseResponse(responseData)
            
            if entries.count == 0 {
                let alertController = UIAlertController(title: NSLocalizedString("Faculty", comment: "title for faculty no match"), message: NSLocalizedString("Person was not found", comment: "Person was not found"), preferredStyle: .Alert)
                let OKAction = UIAlertAction(title: "OK", style: .Default, handler: nil)
                alertController.addAction(OKAction)
                self.presentViewController(alertController, animated: true, completion: nil)
            } else if entries.count == 1 {
                self.performSegueWithIdentifier("Show Faculty Person", sender: entries[0])
            } else {
                self.performSegueWithIdentifier("Show Faculty List", sender: entries)
            }
            
            
        case 2:
            ()
        default:
            ()
        }
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        
        if segue.identifier == "Show Course Location" {
            self.sendEventToTracker1WithCategory(kAnalyticsCategoryUI_Action, withAction: kAnalyticsActionList_Select, withLabel: "Map Detail", withValue: nil, forModuleNamed: self.module!.name)
            let vc = segue.destinationViewController as! POIDetailViewController
            vc.buildingId = sender as? String
            vc.module = self.module
        } else if segue.identifier == "Show Faculty List" {
            let detailController = segue.destinationViewController as! DirectoryViewController
            detailController.entries = sender as! [DirectoryEntry];
            detailController.module = self.module;
        } else if segue.identifier == "Show Faculty Person" {
            let detailController = segue.destinationViewController as! DirectoryEntryViewController
            detailController.entry = sender as? DirectoryEntry;
            detailController.module = self.module;
        }
    }
}
